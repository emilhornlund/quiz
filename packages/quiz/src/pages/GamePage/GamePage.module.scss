@use '../../styles/helpers';
@use '../../styles/variables';

.leaveModalActionButtons {
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  justify-content: center;
  align-items: center;
  align-self: center;
  width: 100%;
  max-width: variables.$mobile-width;

  @include helpers.mobile {
    column-gap: variables.$mobile-control-normal-layout-gutter;
  }

  @include helpers.tablet {
    column-gap: variables.$tablet-control-normal-layout-gutter;
  }

  @include helpers.desktop {
    column-gap: variables.$desktop-control-normal-layout-gutter;
  }

  div[class*="buttonInputContainer"] {
    flex: 1;
  }
}
