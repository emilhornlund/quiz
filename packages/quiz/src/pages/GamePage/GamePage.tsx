import { GameEventType } from '@quiz/common'
import React, { useEffect, useMemo } from 'react'
import { Bounce, toast, ToastOptions } from 'react-toastify'

import { LeaveButton, LoadingSpinner, Page } from '../../components'
import {
  HostGameBeginState,
  HostLeaderboardState,
  HostLobbyState,
  HostPodiumState,
  HostQuestionPreviewState,
  HostQuestionState,
  HostResultState,
  PlayerAwaitingResultState,
  PlayerGameBeginState,
  PlayerLobbyState,
  PlayerPodiumState,
  PlayerQuestionPreviewState,
  PlayerQuestionState,
  PlayerResultState,
} from '../../states'
import {
  ConnectionStatus,
  useEventSource,
} from '../../utils/use-event-source.tsx'
import { useGameTokenQueryParam } from '../../utils/use-game-token-query-param.tsx'
import { useQuizService } from '../../utils/use-quiz-service.tsx'

const GamePage = () => {
  const [token, gameID] = useGameTokenQueryParam()

  const [event, connectionStatus] = useEventSource(gameID, token)

  const { completeTask, submitQuestionAnswer } = useQuizService()

  useEffect(() => {
    if (connectionStatus !== ConnectionStatus.INITIALIZED) {
      const options: ToastOptions = {
        position: 'top-right',
        autoClose: 5000,
        hideProgressBar: false,
        closeOnClick: true,
        pauseOnHover: true,
        draggable: true,
        progress: undefined,
        theme: 'colored',
        transition: Bounce,
      }
      switch (connectionStatus) {
        case 'CONNECTED':
          toast.success('Connected', options)
          break
        case 'RECONNECTING':
          toast.warning('Reconnecting', options)
          break
        case 'RECONNECTING_FAILED':
          toast.error('Reconnecting failed', options)
          break
        default:
          break
      }
    }
  }, [connectionStatus])

  const stateComponent = useMemo(() => {
    switch (event?.type) {
      case GameEventType.GameLobbyHost:
        return (
          <HostLobbyState
            event={event}
            onStart={() => completeTask(gameID!, token!)}
          />
        )
      case GameEventType.GameLobbyPlayer:
        return <PlayerLobbyState event={event} />
      case GameEventType.GameBeginHost:
        return <HostGameBeginState event={event} />
      case GameEventType.GameBeginPlayer:
        return <PlayerGameBeginState event={event} />
      case GameEventType.GameQuestionPreviewHost:
        return <HostQuestionPreviewState event={event} />
      case GameEventType.GameQuestionPreviewPlayer:
        return <PlayerQuestionPreviewState event={event} />
      case GameEventType.GameQuestionHost:
        return <HostQuestionState event={event} />
      case GameEventType.GameQuestionPlayer:
        return (
          <PlayerQuestionState
            event={event}
            onSubmitQuestionAnswer={(request) =>
              submitQuestionAnswer(gameID!, token!, request)
            }
          />
        )
      case GameEventType.GameAwaitingResultPlayer:
        return <PlayerAwaitingResultState event={event} />
      case GameEventType.GameLeaderboardHost:
        return <HostLeaderboardState event={event} />
      case GameEventType.GameResultHost:
        return <HostResultState event={event} />
      case GameEventType.GameResultPlayer:
        return <PlayerResultState event={event} />
      case GameEventType.GamePodiumHost:
        return <HostPodiumState event={event} />
      case GameEventType.GamePodiumPlayer:
        return <PlayerPodiumState event={event} />
      default:
        return (
          <Page header={<LeaveButton />}>
            <LoadingSpinner />
          </Page>
        )
    }
  }, [event, gameID, token, completeTask, submitQuestionAnswer])

  return <>{stateComponent}</>
}

export default GamePage
