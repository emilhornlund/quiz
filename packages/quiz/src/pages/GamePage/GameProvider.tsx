import React, { createContext, FC, ReactNode, useMemo } from 'react'

import { IGameContext } from '../../utils/types.ts'
import { useGameStorage } from '../../utils/use-game-storage.tsx'

export const GameContext = createContext<IGameContext>({})

export interface GameProviderProps {
  children: ReactNode | ReactNode[]
}

export const GameProvider: FC<GameProviderProps> = ({ children }) => {
  const [data] = useGameStorage()

  const contextValue = useMemo<IGameContext>(
    () => ({
      gameID: data.gameID,
      token: data.token,
    }),
    [data],
  )

  return (
    <GameContext.Provider value={contextValue}>{children}</GameContext.Provider>
  )
}
