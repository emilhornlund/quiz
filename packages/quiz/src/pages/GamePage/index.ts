export { default } from './GamePage'
export * from './GameProvider'
