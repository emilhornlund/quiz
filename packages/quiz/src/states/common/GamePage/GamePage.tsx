import { faMaximize, faMinimize } from '@fortawesome/free-solid-svg-icons'
import React, { FC } from 'react'

import { Button, Page, PageProps } from '../../../components'
import { useGameContext } from '../../../context/game'

const GamePage: FC<PageProps> = ({ header, children, ...rest }) => {
  const { isFullscreenActive, toggleFullscreen } = useGameContext()

  return (
    <Page
      header={
        <>
          <Button
            id="fullscreen-button"
            type="button"
            kind="primary"
            size="small"
            icon={isFullscreenActive ? faMinimize : faMaximize}
            onClick={() => toggleFullscreen().then()}
          />
          {header}
        </>
      }
      {...rest}>
      {children}
    </Page>
  )
}

export default GamePage
