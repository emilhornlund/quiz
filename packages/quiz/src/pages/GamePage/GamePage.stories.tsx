import type { Meta, StoryObj } from '@storybook/react'
import { withRouter } from 'storybook-addon-remix-react-router'

import GamePage from './GamePage'

const meta = {
  component: GamePage,
  decorators: [withRouter],
  parameters: {
    layout: 'fullscreen',
  },
} satisfies Meta<typeof GamePage>

export default meta
type Story = StoryObj<typeof meta>

export const Default = {
  args: {},
} satisfies Story
