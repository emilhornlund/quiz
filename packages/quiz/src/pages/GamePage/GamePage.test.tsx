import { render } from '@testing-library/react'
import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { describe, expect, it } from 'vitest'

import GamePage from './GamePage'

describe('GamePage', () => {
  it('should render GamePage', async () => {
    const { container } = render(
      <MemoryRouter>
        <GamePage />
      </MemoryRouter>,
    )

    expect(container).toMatchSnapshot()
  })
})
