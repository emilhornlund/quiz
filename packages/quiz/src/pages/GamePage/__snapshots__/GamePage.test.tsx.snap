// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GamePage > should render GamePage 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Klurigo
        </span>
      </button>
      <div
        class="side"
      >
        <a
          data-discover="true"
          href="/auth/login"
        >
          Login
        </a>
      </div>
    </div>
    <div
      class="content fullWidth centerAlign"
    >
      <div
        class="loadingSpinnerContainer"
      >
        <div />
        <div />
        <div />
      </div>
    </div>
  </div>
</div>
`;
